provisioner/terraform/tfparse/tfparse.go

```diff
@@ -0,0 +1,490 @@
+		}
+		return nil, xerrors.Errorf("provisioner tag %q evaluated to an empty value, please set a default value", k)
+	}
+	return evalTags, nil
+}
+
+func loadModuleFromFile(file []byte, mimetype string) (module *tfconfig.Module, cleanup func() error, err error) {
+	// Create a temporary directory
+	cleanup = func() error { return nil } // no-op cleanup
+	tmpDir, err := os.MkdirTemp("", "tfparse-*")
+	if err != nil {
+		return nil, cleanup, xerrors.Errorf("create temp dir: %w", err)
+	}
+	cleanup = func() error { // real cleanup
+		return os.RemoveAll(tmpDir)
+	}
+
+	// Untar the file into the temporary directory
+	var rdr io.Reader
+	switch mimetype {
+	case "application/x-tar":
+		rdr = bytes.NewReader(file)
+	case "application/zip":
+		zr, err := zip.NewReader(bytes.NewReader(file), int64(len(file)))
+		if err != nil {
+			return nil, cleanup, xerrors.Errorf("read zip file: %w", err)
+		}
+		tarBytes, err := archive.CreateTarFromZip(zr, maxFileSizeBytes)
+		if err != nil {
+			return nil, cleanup, xerrors.Errorf("convert zip to tar: %w", err)
+		}
+		rdr = bytes.NewReader(tarBytes)
+	default:
+		return nil, cleanup, xerrors.Errorf("unsupported mimetype: %s", mimetype)
+	}
+
+	if err := provisionersdk.Untar(tmpDir, rdr); err != nil {
+		return nil, cleanup, xerrors.Errorf("untar: %w", err)
+	}
+
+	module, diags := tfconfig.LoadModule(tmpDir)
+	if diags.HasErrors() {
+		return nil, cleanup, xerrors.Errorf("load module: %s", diags.Error())
+	}
+
+	return module, cleanup, nil
+}
+
+// loadDefaults inspects the given module and returns the default values for
+// all variables and coder_parameter data sources referenced there.
+func loadDefaults(ctx context.Context, logger slog.Logger, module *tfconfig.Module) (varsDefaults map[string]string, paramsDefaults map[string]string, err error) {
+	// iterate through module.Variables to get the default values for all
+	// variables.
+	varsDefaults = make(map[string]string)
+	for _, v := range module.Variables {
+		sv, err := interfaceToString(v.Default)
+		if err != nil {
+			return nil, nil, xerrors.Errorf("can't convert variable default value to string: %v", err)
+		}
+		varsDefaults[v.Name] = strings.Trim(sv, `"`)
+	}
+
+	// iterate through module.DataResources to get the default values for all
+	// coder_parameter data resources.
+	paramsDefaults = make(map[string]string)
+	for _, dataResource := range module.DataResources {
+		if dataResource.Type != "coder_parameter" {
+			logger.Debug(ctx, "skip resource as it is not a coder_parameter", "resource_name", dataResource.Name, "resource_type", dataResource.Type)
+		}
+
+		var file *hcl.File
+		var diags hcl.Diagnostics
+		parser := hclparse.NewParser()
+
+		if !strings.HasSuffix(dataResource.Pos.Filename, ".tf") {
+			logger.Debug(ctx, "only .tf files can be parsed", "filename", dataResource.Pos.Filename)
+			continue
+		}
+
+		// We know in which HCL file is the data resource defined.
+		file, diags = parser.ParseHCLFile(dataResource.Pos.Filename)
```

Comment:
Again, it looks like we re-parse the same file many times?

Reply:
If we want to stick with the idea of raw parsing, I believe we need to implement our own "super parser" that retrieves all data at once. Thoughts?

Reply:
At that point, are we not essentially doing `terraform plan`?

Reply:
Ideally, we should do this but most likely impossible here. Would that require another step in the GRPC handshake (plan vs. provision)?

Reply:
If we do this on the provisioner, how do we figure out the provisioner tags? We get right back to the chicken-egg problem :-)